### ISSUE TITLE
Proposal: Clarification on criteria for stable component status and deprecation of "stable candidate" term.

### ISSUE THREAD

**Issue by berndverst** [dapr/dapr#4970]
/area docs

## Clarification on criteria for stable component designation

As a maintainer of **components-contrib** I propose the following requirements for a component to be designated as stable:

- The component was included in the previous release (both alpha and beta state are acceptable). _This means a brand new component cannot be introduced and immediately be considered stable. It must be at least Alpha or Beta for one release. This of course shall not prevent an eager contributor from contributing a certification test early._
- In addition to conformance tests the component must have a **stable certification test** including a `Readme` (the test plan) outlining what is being tested. The certification test must cover the following requirements:
    1. Every component metadata attribute and their potential values (if this is a list or enum) must be covered by an integration test communicating with the actual cloud service or containerized service for which this component is intended to be used.
    2. Every supported authentication or connection mechanism must be covered by an integration test.
    3. If the component uses a protocol that maintains a connection (e.g. websocket, or binary protocol etc) a test must be created that demonstrates the component is able to reconnect if this connection is interrupted. _Most components communicating via REST or gRPC do not require this. However, components that maintain streams or perform long-polling operations still require such testing._
    4. **Emulators of cloud services are not acceptable for meeting the certification test requirements above.** Certification tests for cloud services must communicate with the actual cloud service. _This may require coordination with the Dapr Steering Community to securely configure credentials that can be used to exercise cloud-based test infrastructure._

_Note: Certification tests should avoid basic operations already covered by conformance tests._

With this new definition above the term **stable candidate** will be dropped. A component may move from **alpha** or **beta** to **stable directly** if the above requirements have been met.

**Comment by berndverst** [comment]
cc @dapr/maintainers-components-contrib @dapr/maintainers-dapr

**Comment by tanvigour** [comment]
This would make certification test stage approval more stricter so that we can directly move beta to stable instead of waiting one release to move to stable, is that correct?

**Comment by berndverst** [comment]
> This would make certification test stage approval more stricter so that we can directly move beta to stable instead of waiting one release to move to stable, is that correct?

No, this formalizes what was already the requirement upheld by contrib maintainers. Further, it removes the unnecessary stable candidate designation which h introduced a waiting period that served no purpose as we haven't seen any increase in adoption of beta components that were stable candidates. This introduced unnecessary complexity and so we are getting rid of this.

In theory with this change a component could be stable faster than before as long as the formalized requirements have been met.

**Comment by yaron2** [comment]
Looks good.

**Comment by halspang** [comment]
Overall looks good to me. Two things:

1. Given the stricter definitions of the cert tests, are we going to go through and mark which ones need to be updated? Certain components like Kafka and Azure Service Bus* have _a lot_ of metadata so I wouldn't be surprised if they were missed.
2. Should we include anything about resiliency here? We have a proposal [1863](https://github.com/dapr/components-contrib/issues/1863) but it hasn't been ratified yet. Do you think it's too early to include what that path would look like?

**Comment by berndverst** [comment]
> Overall looks good to me. Two things:
> 
> 1. Given the stricter definitions of the cert tests, are we going to go through and mark which ones need to be updated? Certain components like Kafka and Azure Service Bus* have _a lot_ of metadata so I wouldn't be surprised if they were missed.
> 2. Should we include anything about resiliency here? We have a proposal [1863](https://github.com/dapr/components-contrib/issues/1863) but it hasn't been ratified yet. Do you think it's too early to include what that path would look like?

1.) Existing components should be grandfathered in. We should make sure to revisit their certification tests when we have an opportunity to do so.

@dapr/approvers-components-contrib @dapr/maintainers-components-contrib should ensure that new metadata options being introduced for stable components also add an appropriate test case in certification tests going forward.

2.) Given that the Component Resiliency Proposal isn't finalized / approved yet we cannot consider it yet as part of this standard.

I do expect that in the future implementing the Resiliency interface will be a requirement for stable certification as well. At that time we will need to amend the published guidelines for stable component certification.

**Comment by yaron2** [comment]
> Existing components should be grandfathered in. We should make sure to revisit their certification tests when we have an opportunity to do so.

> https://github.com/orgs/dapr/teams/approvers-components-contrib https://github.com/orgs/dapr/teams/maintainers-components-contrib should ensure that new metadata options being introduced for stable components also add an appropriate test case in certification tests going forward.

Agree

> Given that the Component Resiliency Proposal isn't finalized / approved yet we cannot consider it yet as part of this standard.

Agree

> I do expect that in the future implementing the Resiliency interface will be a requirement for stable certification as well. 

I'm not too sure about that, but in any case we probably need a few releases with Resiliency in stable status before we can make that decision.

**Comment by ItalyPaleAle** [comment]
Overall I think this makes a lot of sense.

Only one comment on this:

> If the component uses a protocol that maintains a connection (e.g. websocket, or binary protocol etc) a test must be created that demonstrates the component is able to reconnect if this connection is interrupted. Components communicating via REST or gRPC do not require this.

On the last sentence:

1. I think we should clarify that components that use HTTP or GRPC but maintain streams or perform long-polling (which is the case for some pub subs), then they are still required to comply with this requirement.
2. Even when the component does not maintain any persistent connection, I think it is still important to test a situation where the network is down to validate how the component handles that?

**Comment by artursouza** [comment]
Are we OK to merge https://github.com/dapr/docs/pull/2720 ?

**Comment by berndverst** [comment]
> 1. I think we should clarify that components that use HTTP or GRPC but maintain streams or perform long-polling (which is the case for some pub subs), then they are still required to comply with this requirement.

Agreed. Updated.

> 2. Even when the component does not maintain any persistent connection, I think it is still important to test a situation where the network is down to validate how the component handles that?

I don't think that belongs in certification tests. If the network is down the component will fail operations multiple times (depending on resiliency settings). It would be expected that the runtime eventually consider the component to be in bad state and attempts to restart the component (or restarts the entire sidecar -- whichever the current behavior is).

**Comment by berndverst** [comment]
> Are we OK to merge [dapr/docs#2720](https://github.com/dapr/docs/pull/2720) ?

Please update the PR to also add the detailed requirements on the certification tests which we can link to. Otherwise the page is still very ambiguous as it isn't clear what a certification test entails.
